### ISSUE TITLE
Distinguishing between webgme/webgme and webgme/webgme-cli

### ISSUE THREAD

**Issue by sprinkjm** [webgme/webgme#1220]
I have some confusion between webgme and webgme-cli for a "production" environment. In my course I would like to have a common server where students could login (I have authentication working now) and
- develop their own languages
- develop their own plugins/decorators
- interpret their models (as well as models I might provide)
  While WebGME may not support all this now, I'm trying to figure out to run component generators in a webgme environment (e.g., where I run `npm start` not `webgme start`).

So in summary: if I create a language using webgme on a web server, how can I create a plugin/addon for that language using the plugin-generator or the webgme-cli tools, without having created a project and run a separate webgme server for that project? Or is there a different mechanism I should use?

**Comment by kecso** [comment]
We prefer the self contained webgme-cli based method for developing plugins or visualizers as they allow easy debug and  - by being self-contained in a github repository - easier portability.

**The option(s) you have if you would like to run a single deployment/server:**
- You will need a separate webserver that will allow students to upload their plugins/decorators (within this server you probably be able to use the authentication methods provided by webgme so you do not have to build a separate database for that purpose)
- You would also need a watchdog that would restart your webgme server every time some content is uploaded.
- Additionally you just need two directories where you would store these addons (one for plugins, an other for decorators), and you need to configure them properly int your [gmeConfig](https://github.com/webgme/webgme/blob/master/config/README.md).

**A more integrated option would be having the following webgme-cli based project:**
- You create a single webgme-cli project, import all provided addons(plugins and decorators), that you need or want to provide to your students.
- You add/implement a 'router' which basically provides additional web-pages where you can solve the upload/download, creation/deletion of the addons that will be created by the students (basically you transform the web-input of students into either webgme-cli commands or plain file-system operations)
- That will save you the hassle of manipulating gmeconfig yourself.
- You will still need some watchdog (or preferably nodemon) process that would be able to initiate server restart whenever the additional content changes (as long as the students are connected to the models, they will be reconnected, but some upload/download kind of operation can fail, so you have to make sure that the users are notified properly)

**Comment by kecso** [comment]
About the porting of generated plugins:
- if you generate (using ```webgme new plugin``` command) your plugin with webgme-cli, you can port it by copying the whole PluginName (under src/plugins directory) directory with its content to one of your configured pluginBasePaths directory
- if generating with the PluginGenerator plugin, then you need to do the same with the directory PluginName (inside directory src/plugins/projectName/PluginName) and its contents

**Comment by sprinkjm** [comment]
Can you help me understand how these options might change if I have say 20+ students, and each has their own language they develop and in which they develop models? I guess this is my main question: how do I run webgme if I have more than one "Project"? 

I see in some cases this sequence of events:

1. Install webgme and run `npm start`
2. Visit localhost:8888 and create MyNewProject and MySecondNewProject

In other cases it is:

1. Install webgme-cli and run `webgme init MyNewProject`
2. `cd MyNewProject`
3. `webgme start`
4. `webgme new plugin MyNewPlugin MyNewProject`

In the first case, I can create arbitrarily many projects and they can be assigned/available to arbitrarily many users. What I don't know how to do is the equivalent of run `webgme new plugin` in order to create a plugin for a project that exists as created on this instance of webgme. How would I enable a user to create a plugin /only/ for their project that they created through the web interface?

In the second case, I don't know how to have more than one project that is available. All projects created from the first case are not available (presumably because the webgme instance is installed and run from a different directory). I am hoping to have 20+ versions of projects available at different times, to 20+ different users, where each user should have access only to their project and be able to run and debug their own plugins on the common server.

So each of these has a limitation I cannot overcome, can you help me find a way to disambiguate and solve them? Or is the server infrax meant only for stable projects with a single system admin who administers the directories behind the scenes?

**Comment by kecso** [comment]
First of all, let's clear the unfortunate overloading of the term 'project'.
If you use webgme-cli, then you create webgmeApps not project (I usually make that mistake as well). The wbgmeApp couples plugins, decorators, visualizers, and project seeds.
The project is the domain+model structure you create using WebGME. By itself it holds only information necessary to model interpretation, but the additional components are not coupled with it.
This means that there is no way to limit a plugin to a certain project.

Also in your second scenario, when you create/run/develop a webgmeApp, you are still able to create arbitrary number of projects and they will actually have no limitations or any hard-wired conneciton to the webgmeApp.


## An example setup for this kind of scenario:
### The student's deployment:
- The students strictly use the webgme-cli, create a webgmeApp
- They develop their components inside that webgmeApp
- They create their projects with the help of that webgmeApp
- They save their webgmeApp into a github repository
- They export their project(s) using the standard export functionality of their deployment

### The common deployment:
- Create a webgmeApp using webgme-cli
- Run this deployment so the students can create projects
- Whenever a student submits his/her work (in the form of the github repository) you can make their components available by importing them with the help of webgme-cli
- In case of updates, you can execute an npm install or 're-import' the components
- All student will have an account on this deployment (with the 'canCreate' flag)
- Once they submitted their work, they can create their project(s) by importing their exported 'webgmex' sources
- They can also grant access to other users (like the instructor) so they will be able to see their work, but the default behavior is that only they can access their uploaded project(s)

**Comment by kecso** [comment]
For model aware decoration and model interpretation inside UI see:
- [dynamic SVGs](https://www.youtube.com/watch?v=l5m4CF4w8fE)
- [iCore](https://www.youtube.com/watch?v=gDvLnR0iDJQ)
